server/auth/types/openid/routes.ts

```diff
@@ -271,5 +284,116 @@ export class OpenIdAuthRoutes {
+          <link rel="icon" href="data:,">
+          <script src="${serverBasePath}/auth/openid/captureUrlFragment.js"></script>
+        `,
+        });
+      }
+    );
+
+    // This script will store the URL Hash in browser's local storage.
+    this.core.http.resources.register(
+      {
+        path: '/auth/openid/captureUrlFragment.js',
+        validate: false,
+        options: {
+          authRequired: false,
+        },
+      },
+      async (context, request, response) => {
+        this.sessionStorageFactory.asScoped(request).clear();
+        return response.renderJs({
+          body: `let oidcHash=window.location.hash.toString();
+                let redirectHash = false;
+                if (oidcHash !== "") {
+                  window.localStorage.removeItem('oidcHash');
+                  window.localStorage.setItem('oidcHash', oidcHash);
+                    redirectHash = true;
+                }
+                let params = new URLSearchParams(window.location.search);
+                let nextUrl = params.get("nextUrl");
+                finalUrl = "login?nextUrl=" + encodeURIComponent(nextUrl);
+                finalUrl += "&redirectHash=" + encodeURIComponent(redirectHash);
+                window.location.replace(finalUrl);
+              `,
+        });
+      }
+    );
+
+    //  Once the User is authenticated the browser will be redirected to '/auth/openid/redirectUrlFragment'
+    //  route, which will execute the redirectUrlFragment.js.
+    this.core.http.resources.register(
+      {
+        path: '/auth/openid/redirectUrlFragment',
+        validate: {
+          query: schema.object({
+            nextUrl: schema.any(),
+          }),
+        },
+        options: {
+          authRequired: true,
+        },
+      },
+      async (context, request, response) => {
+        const serverBasePath = this.core.http.basePath.serverBasePath;
+        return response.renderHtml({
+          body: `
+          <!DOCTYPE html>
+          <title>OSD OpenID Success</title>
+          <link rel="icon" href="data:,">
+          <script src="${serverBasePath}/auth/openid/redirectUrlFragment.js"></script>
+        `,
+        });
+      }
+    );
+
+    // This script will pop the Hash from local storage if it exists.
+    // And forward the browser to the next url.
+    this.core.http.resources.register(
+      {
+        path: '/auth/openid/redirectUrlFragment.js',
+        validate: false,
+        options: {
+          authRequired: true,
+        },
+      },
+      async (context, request, response) => {
+        return response.renderJs({
+          body: `let oidcHash=window.localStorage.getItem('oidcHash');
+                window.localStorage.removeItem('oidcHash');
+                let params = new URLSearchParams(window.location.search);
+                let nextUrl = params.get("nextUrl");
+                finalUrl = nextUrl + oidcHash;
+                window.location.replace(finalUrl);
```

Comment:
Leaving a comment here for myself, want to test this, can we use core's redirect somewhere if we still have access to it, or we only have access to JS functions at this point?